### PR TITLE
Copy name of outputs in multi_gpu_model.

### PR DESCRIPTION
**Pull request by hgaiser**
This PR copies the names of the outputs to the multi_gpu_model output. This allows the use of loss dictionaries, where the name of the output is coupled with a loss function.

ps. I tried to do the same for the input, but it seems that `model.input_names` is not yet created at that time (it is created in the line with `outputs = model(inputs)`). Got any tips to fix this? My issue was with outputs, but it seems other people have the same problem with the inputs (https://github.com/fchollet/keras/issues/8339).

### CODE REVIEW

**Comment by fchollet** [comment]
> but it seems that model.input_names is not yet created at that time (it is created in the line with outputs = model(inputs)

Not sure I follow; all models have an `input_names` attribute (since it is set by the model's constructor): https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1620-L1635

You can't instantiate a model without it, so the model passed to `multi_gpu_model` will definitely have it.

File: keras/utils/training_utils.py [hunk 1]
```diff
@@ -155,13 +153,13 @@ def get_slice(data, i, parts):
...
-                for o in range(len(outputs)):
-                    all_outputs[o].append(outputs[o])
+                for name, o in zip(model.output_names, outputs):
+                    all_outputs.setdefault(name, []).append(o)
 
     # Merge outputs on CPU.
     with tf.device('/cpu:0'):
         merged = []
-        for outputs in all_outputs:
+        for name, outputs in all_outputs.items():
```
Comment (fchollet) [inline keras/utils/training_utils.py]: This is no longer ordered, which I think is not intuitive.

Simpler: don't change anything until this line (i.e. use a list of lists), and at this line, use `zip(model.output_names, all_outputs)`.

Comment (hgaiser) [reply keras/utils/training_utils.py]: Good point.

**Comment by hgaiser** [comment]
> Not sure I follow; all models have an input_names attribute (since it is set by the model's constructor): https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1620-L1635
> 
> You can't instantiate a model without it, so the model passed to multi_gpu_model will definitely have it.

In the case of the functional API, you are correct. However I was testing this on `mnist_cnn.py` which uses a `Sequential` model. A `Sequential` model's [`__init__`](https://github.com/fchollet/keras/blob/master/keras/models.py#L386-L408) does not call its `super.__init__` so the `__init__` of `Container` (which sets the input names as you mentioned) is never called.

**Comment by fchollet** [comment]
Then you can do a typecheck for `Sequential`, and in that case use `model.model.input_names`, since that is equivalent: https://github.com/fchollet/keras/blob/master/keras/models.py#L571

Please add input name support in this PR.

**Comment by hgaiser** [comment]
Done. I do have to build the `Sequential` model first, otherwise the `input_names` just doesn't exist.
